signature/src/mechanism/cosign/mod.rs

```diff
@@ -0,0 +1,346 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//! Cosign verification
+
+use std::{collections::HashMap, path::Path, str::FromStr};
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use sigstore_rs::{
+    cosign::{
+        verification_constraint::{PublicKeyVerifier, VerificationConstraintVec},
+        verify_constraints, CosignCapabilities,
+    },
+    crypto::SignatureDigestAlgorithm,
+    errors::SigstoreVerifyConstraintsError,
+};
+use tokio::fs;
+use url::Url;
+
+use crate::{
+    mechanism::{Image, SignScheme},
+    payload::simple_signing::SigPayload,
+    policy::ref_match::PolicyReqMatchType,
+};
+
+/// Dir for storage of cosign verification keys.
+pub const COSIGN_KEY_DIR: &str = "/run/image-security/cosign";
+
+/// The name of resource to request cosign verification key from kbs
+pub const COSIGN_KEY_KBS: &str = "Cosign Key";
+
+#[derive(Deserialize, Debug, Eq, PartialEq, Serialize, Default)]
+pub struct CosignParameters {
+    // KeyPath is a pathname to a local file containing the trusted key(s).
+    // Exactly one of KeyPath and KeyData can be specified.
+    //
+    // This field is optional.
+    #[serde(rename = "keyPath")]
+    pub key_path: Option<String>,
+    // KeyData contains the trusted key(s), base64-encoded.
+    // Exactly one of KeyPath and KeyData can be specified.
+    //
+    // This field is optional.
+    #[serde(rename = "keyData")]
+    pub key_data: Option<String>,
+
+    // SignedIdentity specifies what image identity the signature must be claiming about the image.
+    // Defaults to "match-exact" if not specified.
+    //
+    // This field is optional.
+    #[serde(default, rename = "signedIdentity")]
+    pub signed_identity: Option<PolicyReqMatchType>,
+}
+
+#[async_trait]
+impl SignScheme for CosignParameters {
+    /// This initialization will:
+    /// * Create [`COSIGN_KEY_DIR`] if not exist.
+    async fn init(&self) -> Result<()> {
+        if !Path::new(COSIGN_KEY_DIR).exists() {
+            fs::create_dir_all(COSIGN_KEY_DIR).await.map_err(|e| {
+                anyhow!("Create Simple Signing sigstore-config dir failed: {:?}", e)
+            })?;
+        }
+
+        Ok(())
+    }
+
+    fn resource_manifest(&self) -> HashMap<&str, &str> {
+        let mut manifest = HashMap::new();
+        if let Some(key_path) = &self.key_path {
+            if !Path::new(key_path).exists() {
+                manifest.insert(COSIGN_KEY_KBS, &key_path[..]);
```

Comment:
I'm rather confused by this. If the key doesn't exist, we add it's path. But what if the file does exist? But also we're effectively doing `insert("key", path)`, which might be confusing as you may well expect the code to do `insert("key", key)`.

Reply:
I'm not sure what confused you. Here `key_path` means `path-to-the-key`, e.g., `/etc/keys/cosign.pub`. It's not `/etc/keys`.

So I think here may be right. If there is no file `/etc/keys/cosign.pub`, the hashmap will be inserted `("Cosign Key", "/etc/keys/cosign.pub")`. Here `"Cosign Key"` tells the KBC to resource name to be requested. `"/etc/keys/cosign.pub"` declares when the resource bit stream is received, it will be stored into `"/etc/keys/cosign.pub"`

Reply:
maybe change the Hashmap name from `manifest` to `manifest_from_kbs` can be less confusing?